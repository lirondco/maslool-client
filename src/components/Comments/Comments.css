.trail_comment {
  margin-bottom: 2%;
  padding: 1%;
  border: #524c57 solid;
  border-radius: 10px;
  box-shadow: 5px 5px 2.5px grey;
}

.edit_comment {
  margin: 0 !important;
}

.comment_buttons {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

.comment_buttons button {
  margin: 0 !important;
  border-radius: 10px;
}

.comment_buttons button:hover {
  background-color: #fffcf5;
  border-color: #999;
}

.flag_button {
  margin: 0 !important;
  border: none;
  color: #900909;
  text-decoration: underline;
}

.flag_button:hover {
  margin: 0 !important;
  border: none;
  color: maroon;
  text-decoration: underline;
  font-weight: bold;
}

.flagged_comment {
  color: #900909;
}

.comment_date {
  font-style: italic;
}

.comment_content {
  padding: 2%;
  background-color: white;
  border: #524c57 solid 1px;
  border-radius: 10px;
  margin: 1%;
}
