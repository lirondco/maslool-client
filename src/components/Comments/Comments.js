import React, { Component } from "react";
import { NiceDate, Textarea } from "../../components/Utils/Utils";
import UserContext from "../../contexts/UserContext";
import TrailsApiService from "../../services/trails-api-service";
import "./Comments.css";

export default class Comments extends Component {
  state = { commentEdited: null };
  static defaultProps = {
    onFlagSuccess: () => {},
    onEditSuccess: () => {},
    onDeleteSuccess: () => {},
  };

  static contextType = UserContext;

  handleEditActiveClick = (comment, ev) => {
    ev.preventDefault();
    this.setState({ commentEdited: comment.id });
  };

  handleCancelEditClick = (ev) => {
    ev.preventDefault();
    this.setState({ commentEdited: null });
  };

  renderButtons = (comment) => {
    return (
      <div className="comment_buttons">
        <button
          className="edit_comment_button"
          onClick={(e) => this.handleEditActiveClick(comment, e)}
        >
          Edit
        </button>
        <button
          className="delete_comment_button"
          onClick={(e) => this.handleDelete(comment.id, e)}
        >
          Delete
        </button>
      </div>
    );
  };

  renderEditButtons = () => {
    return (
      <div className="comment_buttons">
        <button type="submit" className="edit_comment_button">
          Submit Edit
        </button>
        <button
          className="edit_comment_button"
          onClick={this.handleCancelEditClick}
        >
          Cancel
        </button>
      </div>
    );
  };

  handleFlaggedClick = (comment, ev) => {
    const { user } = this.context;
    if (comment.user.id === user.id) {
      alert("You cannot flag your own comment!");
      return;
    }
    ev.preventDefault();
    this.context.clearError();
    TrailsApiService.flagComment(comment.id)
      .then(this.props.onFlagSuccess())
      .catch(this.context.setError);
  };

  handleEditSubmit = (comment, ev) => {
    ev.preventDefault();
    this.context.clearError();
    const { content } = ev.target;
    TrailsApiService.editComment(comment.id, content.value)
      .then(this.setState({ commentEdited: null }))
      .then(() => {
        content.value = "";
      })
      .then(this.props.onEditSuccess())
      .catch(this.context.setError);
  };

  handleDelete = (commentId, ev) => {
    ev.preventDefault();
    this.context.clearError();
    TrailsApiService.deleteComment(commentId)
      .then(this.props.onDeleteSuccess())
      .catch(this.context.setError);
  };

  render() {
    const comment = this.props.comment;
    const { user } = this.context;
    let renderedContent;
    if (this.state.commentEdited === comment.id) {
      renderedContent = (
        <>
          <form
            className="edit_comment"
            type="submit"
            onSubmit={(e) => this.handleEditSubmit(comment, e)}
          >
            <div className="comment-content">
              <Textarea
                required
                aria-label="Edit your comment..."
                name="content"
                id="content"
                defaultValue={`${comment.content}`}
              />
              {this.renderEditButtons()}
            </div>
          </form>
        </>
      );
    } else {
      renderedContent = (
        <>
          {user.id === comment.user.id || user.admin === true
            ? this.renderButtons(comment)
            : ""}
          <p className="comment_content" style={{whiteSpace: "pre-wrap"}}>{comment.content}</p>
          {comment.flagged === true ? (
            <>
              <p className="flagged_comment">
                Comment has been flagged as inappropriate
              </p>{" "}
              {user.admin === true ? (
                <button
                  className="flag_button"
                  onClick={(e) => this.handleFlaggedClick(comment, e)}
                >
                  Unflag
                </button>
              ) : (
                ""
              )}
            </>
          ) : (
            <button
              className="flag_button"
              onClick={(e) => this.handleFlaggedClick(comment, e)}
            >
              Flag as inappropriate
            </button>
          )}
        </>
      );
    }
    return (
      <div className="trail_comment">
        <h4 className="comment_username">
          From: {comment.user.username} {comment.user.banned ? "[banned]" : ""}
          {comment.user.admin ? "[moderator]" : ""}
        </h4>
        <p className="comment_date">
          Last modified: <NiceDate date={comment.last_modified} />
        </p>
        {renderedContent}
      </div>
    );
  }
}
