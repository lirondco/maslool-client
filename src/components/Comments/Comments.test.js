import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Comments from './Comments'
import { withContext } from 'shallow-with-context'


describe(`Comments component`, () => {
    const props =
        {
            id: 9,
            content: "test comment test comment",
            last_modified: "2021-03-18T15:30:29.873Z",
            flagged: true,
            flaged_by: null,
            user: {
                id: 3,
                username: "test_user",
                join_date: "2021-03-11T06:47:06.477Z",
                admin: false,
                banned: false
            }
        }

        const context = { user: { id: 3,
            admin: false,
            banned: false,
            username: 'test_user' }}
        const ComponentWithContext = withContext(Comments, context)

        it('renders a comment', () => {
            const wrapper = shallow(<ComponentWithContext comment={props} />, { context })
            expect(toJson(wrapper)).toMatchSnapshot()
        })

        it('button works when clicked', () => {
            const wrapper = shallow(<ComponentWithContext comment={props} />, { context })
            wrapper.find('button').at(0).simulate('click', { preventDefault() {}})
        })
})