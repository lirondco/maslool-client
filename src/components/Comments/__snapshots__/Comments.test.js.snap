// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments component renders a comment 1`] = `
<div
  className="trail_comment"
>
  <h4
    className="comment_username"
  >
    From: 
    test_user
     
  </h4>
  <p
    className="comment_date"
  >
    Last modified: 
    <NiceDate
      date="2021-03-18T15:30:29.873Z"
    />
  </p>
  <div
    className="comment_buttons"
  >
    <button
      className="edit_comment_button"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="delete_comment_button"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
  <p
    className="comment_content"
    style={
      Object {
        "whiteSpace": "pre-wrap",
      }
    }
  >
    test comment test comment
  </p>
  <p
    className="flagged_comment"
  >
    Comment has been flagged as inappropriate
  </p>
   
</div>
`;
